src/tcgwars/logic/impl/gen4/DiamondPearlPromos.groovy

```diff
@@ -291,248 +202,243 @@ public enum DiamondPearlPromos implements LogicCardInfo {
+              }
             }
           }
-
         };
       case DRIFBLIM_DP34:
         return copy(DiamondPearl.DRIFBLIM_24, this);
       case PORYGON_Z_DP35:
         return evolution (this, from:"Porygon2", hp:HP120, type:COLORLESS, retreatCost:2) {
           weakness F, PLUS30
           pokePower "Install", {
-            text "As often as you like during your turn , move a Technical Machine card attached to 1 of your Pok�mon to another of your Pok�mon. This power can�t be used if Porygon-Z is affected by a Special Condition."
+            text "As often as you like during your turn (before your attack), move a Technical Machine card attached to 1 of your Pokémon to another of your Pokémon. This power can't be used if Porygon-Z is affected by a Special Condition."
             actionA {
+              checkNoSPC()
+              assert my.all.findAll {it.cards.hasType(TECHNICAL_MACHINE)} : "No Technical Machine cards attached to your Pokémon"
+              assert my.all.size() >= 2 : "Need at least 2 Pokémon"
+              powerUsed()
+
+              def src = my.all.findAll {it.cards.hasType(TECHNICAL_MACHINE)}.select("Source for Technical Machine")
+              def technicalMachine = src.cards.filterByType(TECHNICAL_MACHINE).select("Card to move").first()
+              def tar = my.all
+              tar.remove(src)
+              tar = tar.select("Target for $technicalMachine")
+
+              if (technicalMachine && tar) {
+                src.cards.remove(technicalMachine)
+                tar.cards.add(technicalMachine)
+                bc "$thisAbility moved $technicalMachine from $src to $tar"
+              }
             }
           }
           move "Learning", {
-            text "that levels up from 1 of your Pok�mon, and put it onto that Pok�mon. (This counts as leveling up that Pok�mon.) Shuffle your deck afterward."
+            text "Search your deck for a Pokémon LV.X that levels up from 1 of your Pokémon and put it onto that Pokémon. (This counts as leveling up that Pokémon). Shuffle your deck afterward."
             energyCost ()
-            attackRequirement {}
+            attackRequirement {
+              assert my.deck : "Deck is empty"
+            }
             onAttack {
-              damage 0
+              def predecessors = []
+              my.all.each {
+                predecessors.add(it.name)
+              }
+              def lvX = my.deck.search("Select a Pokémon LV.X that evolves from one of your Pokémon", {
+                it.cardTypes.is(LVL_X) && it.predecessor in predecessors
+              }).first()
+
+              if (lvX) {
+                def target = my.all.findAll {
+                  it.topPokemonCard.name == (lvX as LevelUpPokemonCard).predecessor
+                }.select("Which Pokémon to Level Up?")
+
+                bg().em().run(new LevelUp(target, lvX));
+                bg().em().run(new ActivateAbilities((PokemonCard) lvX, target, reason));
```

Comment:
What is the purpose of this? Shouldn't running the levelUp activate the abilities by default? Also reason isn't defined.